File: tetrad-lib/src/main/java/edu/cmu/tetrad/search/Glasso.java
///////////////////////////////////////////////////////////////////////////////
// For information as to what this class does, see the Javadoc, below.       //
// Copyright (C) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006,       //
// 2007, 2008, 2009, 2010, 2014, 2015 by Peter Spirtes, Richard Scheines, Joseph   //
// Ramsey, and Clark Glymour.                                                //
//                                                                           //
// This program is free software; you can redistribute it and/or modify      //
// it under the terms of the GNU General Public License as published by      //
// the Free Software Foundation; either version 2 of the License, or         //
// (at your option) any later version.                                       //
//                                                                           //
// This program is distributed in the hope that it will be useful,           //
// but WITHOUT ANY WARRANTY; without even the implied warranty of            //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             //
// GNU General Public License for more details.                              //
//                                                                           //
// You should have received a copy of the GNU General Public License         //
// along with this program; if not, write to the Free Software               //
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA //
///////////////////////////////////////////////////////////////////////////////

package edu.cmu.tetrad.search;

import cern.colt.matrix.DoubleMatrix1D;
import cern.colt.matrix.DoubleMatrix2D;
import cern.colt.matrix.impl.DenseDoubleMatrix1D;
import cern.colt.matrix.impl.DenseDoubleMatrix2D;
import cern.colt.matrix.linalg.Algebra;
import cern.jet.math.Mult;
import cern.jet.math.PlusMult;

/**
 * A translation from Tibshirani's 2008 Fortran implementation of glasso.
 *
 * @author Joseph Ramsey
 */
public class Glasso {

    /**
     * Dimension of matrix.
     */
    private int n;

    /**
     * Data covariance matrix.
     */
    private DoubleMatrix2D ss;

    /**
     * Regularization strength parameters for each element (must be symmetric, rho(i, j) = rho(j, i).
     * False by default.
     */
    private Rho rho = (i, j) -> 0;


    /**
     * Maximum number of iterations (no effect if ia = true).
     */
    private int maxit = 10000;


    /**
     * Approximation flag. False if exact solution, true if Meinhausen-Buhlman approximation.
     * False by default.
     */
    private boolean ia;

    /**
     * Initialization flag. false if cold start, initialize using ss. True if warm start, initialize with
     * previous solution stored in ww and wwi. False by default.
     */
    private boolean is;

    /**
     * Trace flag. True if trace information printed. False if trace information not printed.
     * False by default.
     */
    private boolean itr;

    /**
     * Diagonal penalty flag. True if diagonal is penalized. False if diagonal is not penalized.
     * False by default.
     */
    private boolean ipen;

    /**
     * Convergence threshold: Iterations stop when absolute average parameter change is less than
     * thr * avg(abs(offdiagonal(ss)). (Suggested default 1.0e-4.)
     */
    private double thr = 1.0e-4;

    /**
     * Return value of the algorithm.
     */
    public static class Result {

        /**
         * solution inverse covariance matrix estimate (ia = 0)
         * = off-diagonal lasso coefficients (ia != 0)
         */
        private final DoubleMatrix2D wwi;

        /**
         * number of iterations
         */
        private final int niter;

        public Result(DoubleMatrix2D wwi, int niter) {
            this.wwi = wwi;
            this.niter = niter;
        }

        public DoubleMatrix2D getWwi() {
            return this.wwi;
        }

        public int getNiter() {
            return this.niter;
        }

    }

    public Glasso(DoubleMatrix2D cov) {

        this.n = cov.rows();
        this.ss = cov;
    }

    public Result search() {
        int niter = 0;
        final double eps = 1.0e-7;
        int n = getN();
        DoubleMatrix2D ss = getSs();

        boolean approximateAlgorithm = isIa();
        boolean warmStart = isIs();
        boolean itr = isItr();
        boolean pen = isIpen();

//        System.out.println(ss);

        Rho rho = getRho();
        DoubleMatrix2D ww = new DenseDoubleMatrix2D(n, n);
        DoubleMatrix2D wwi = new DenseDoubleMatrix2D(n, n);

        double dlx;

        int nm1 = n - 1;

        DoubleMatrix2D vv = new DenseDoubleMatrix2D(nm1, nm1);
        DoubleMatrix2D xs = null;
        if (!approximateAlgorithm) {
            xs = new DenseDoubleMatrix2D(nm1, n);
        }
        DoubleMatrix1D s = new DenseDoubleMatrix1D(nm1);
        DoubleMatrix1D so = new DenseDoubleMatrix1D(nm1);
        DoubleMatrix1D x = new DenseDoubleMatrix1D(n - 1);
        DoubleMatrix1D ws;
        int[] mm = new int[nm1];
        DoubleMatrix1D ro = new DenseDoubleMatrix1D(nm1);

        // shr warmStart sum(abs(offdiagonal(ss))).
        double shr = 0.0;

        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                if (j == k) continue;
                shr += Math.abs(ss.get(j, k));
            }
        }


        // TO TEST THE FOLLOWING CODE I NEED A DIAGONAL COVARIANCE MATRIX.

        // If ss is diagonal, just return the inverse of the covariance matrix diagonal
        // (penalized if necessary).
        if (shr == 0.0) {
            for (int j = 0; j < n; j++) {
                if (Thread.currentThread().isInterrupted()) {
                    break;
                }

                if (!pen) {
                    ww.set(j, j, ss.get(j, j));
                } else {
                    ww.set(j, j, ss.get(j, j) + rho.get(j, j));
                }
                wwi.set(j, j, 1.0 / Math.max(ww.get(j, j), eps));
            }
            return new Result(wwi, niter);
        }


        shr = getThr() * shr / nm1;

        if (approximateAlgorithm) {
            if (!warmStart) {
                zero(wwi);
            }

            for (int m = 0; m < n; m++) {
                System.out.println("m = " + m);

                // This sets up vv, s, and r--i.e., W.11, s.12, and r.12.
                setup(m, n, ss, rho, ss, vv, s, ro);

                // This sets up x.12--i.e. theta.12.
                int l = -1;

                for (int j = 0; j < n; j++) {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }

                    if (j == m) continue;
                    l = l + 1;
                    x.set(l, wwi.get(j, m));
                }

                lasso(ro, nm1, vv, s, shr / n, x, mm);

                l = -1;
                for (int j = 0; j < n; j++) {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }

                    if (j == m) continue;
                    l = l + 1;
                    wwi.set(j, m, x.get(l));
                }
            }


            niter = 1;
            return new Result(wwi, niter);
        }

//        System.out.println("wwi = " + wwi);

        if (!warmStart) {
            ww.assign(ss);

            zero(xs);
        } else {
            for (int j = 0; j < n; j++) {
                double xjj = -wwi.get(j, j);
//                System.out.println("xjj = " + xjj);
                int l = -1;

                for (int k = 0; k < n; k++) {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }

                    if (k == j) continue;
                    l = l + 1;
                    xs.set(l, j, wwi.get(k, j) / xjj);
                }
            }
        }

        for (int j = 0; j < n; j++) {
            if (pen) {
                ww.set(j, j, ss.get(j, j) + rho.get(j, j));
            } else {
                ww.set(j, j, ss.get(j, j));
//                System.out.println(ww);
            }
        }


        while (true) {
            dlx = 0.0;

            for (int m = 0; m < n; m++) {
                if (itr) {
                    System.out.println("Outer loop = " + m);
                }

                x = xs.viewColumn(m);

                ws = ww.viewColumn(m);

                // This sets up vv, s, and ro--i.e., W.11, s.12, and r.12.
                setup(m, n, ss, rho, ww, vv, s, ro);


                so.assign(s);

//                System.out.println("ww = " + ww);

                // This updates s and x--the estimated correlation matrix and the reduced form of the
                // estimated inverse covariance.
                lasso(ro, nm1, vv, s, shr / sum_abs(vv), x, mm);
//                lasso(ro,nm1,vv,s,thr/sum_abs(vv),x,z,mm);
                int l = -1;

                for (int j = 0; j < n; j++) {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }

                    if (j == m) continue;
                    l = l + 1;
                    ww.set(j, m, so.get(l) - s.get(l));
                    ww.set(m, j, ww.get(j, m));
                }

                dlx = Math.max(dlx, sum_abs_diff(ww.viewColumn(m), ws));
//                xs(:,m)=x
                xs.viewColumn(m).assign(x);
            }

            niter = niter + 1;
            if (niter < getMaxit()) break;
            if (dlx < shr) break;
        }

        inv(n, ww, xs, wwi);

        return new Result(wwi, niter);
    }

    private double sum_abs(DoubleMatrix2D m) {
        double sum = 0.0;

        for (int i = 0; i < m.rows(); i++) {
            for (int j = 0; j < m.columns(); j++) {
                sum += Math.abs(m.get(i, j));
            }
        }

        return sum;
    }

    private double sum_abs_diff(DoubleMatrix1D x, DoubleMatrix1D y) {
        double sum = 0.0;

        for (int i = 0; i < x.size(); i++) {
            sum += Math.abs(x.get(i) - y.get(i));
        }

        return sum;
    }

    private void setup(int m, int n, DoubleMatrix2D ss, Rho rho, DoubleMatrix2D ww, DoubleMatrix2D vv,
                       DoubleMatrix1D s, DoubleMatrix1D r) {
        int l = -1;

        for (int j = 0; j < n; j++) {
            if (j == m) continue;
            l = l + 1;
            r.set(l, rho.get(j, m));                // r is r12
            s.set(l, ss.get(j, m));                 // s is s12
            int i = -1;
            for (int k = 0; k < n; k++) {
                if (Thread.currentThread().isInterrupted()) {
                    break;
                }

                if (k == m) continue;
                i++;
                vv.set(i, l, ww.get(k, j));         // vv is ww.11
            }
        }
    }

    private void lasso(DoubleMatrix1D ro, int n, DoubleMatrix2D vv, DoubleMatrix1D s, double thr,
                       DoubleMatrix1D x, int[] mm) {
        // vv = W.11
        // s = s.12
        // ro = r.12
        // x = x.12 = theta.12

        // z is just passed in to avoid reallocating an array. It's only used in fatmul. Could be global.

        // s = vv * x, or s12 = Theta.1 * Theta.12
        fatmul(n, vv, x, s, mm);

        while (true) {

            // The maximal difference of the beta update of v(j) and v(j).
            double dlx = 0.0;

            for (int j = 0; j < n; j++) {
                double xj = x.get(j);
                x.set(j, 0.0);

                // There is no sum. In the paper there is a sum. Also, there is a minus instead of
                // a plus in the paper.
                double t = s.get(j) + vv.get(j, j) * xj;

                if (Math.abs(t) - ro.get(j) > 0.0) {
                    x.set(j, Math.signum(t) * (Math.abs(t) - ro.get(j)) / vv.get(j, j));
                }

                if (x.get(j) == xj) continue;
                double del = x.get(j) - xj;
                dlx = Math.max(dlx, Math.abs(del));

                for (int i = 0; i < s.size(); i++) {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }

                    s.set(i, s.get(i) - del * vv.get(i, j));
                }
            }

//            System.out.println("dlx = " + dlx + " thr = " + thr);

            if (dlx < thr) break;
        }
    }

    // s = vv * x, or s12 = Theta.1 * Theta.12
    private void fatmul(int n, DoubleMatrix2D vv, DoubleMatrix1D x, DoubleMatrix1D s,
                        int[] m) {
        final double fac = 0.2;

        // z consists of the nonzero entries of x. m indexes these. If there are enough zeroes in x,
        // use the simpler multiplication method.
        int l = 0;

        for (int j = 0; j < n; j++) {
            if (x.get(j) == 0.0) continue;
            l = l + 1;
            m[l] = j;
        }

        if (l < (int) (fac * n)) {
            for (int j = 0; j < n; j++) {
                if (Thread.currentThread().isInterrupted()) {
                    break;
                }

                double dotProduct = 0.0;

                for (int i = 0; i < l; i++) {
                    dotProduct += vv.get(m[i], j) * x.get(m[j]);
                }

                s.set(j, s.get(j) - dotProduct);
            }
        } else {
            s.assign(new Algebra().mult(vv, x), PlusMult.plusMult(-1));
        }

    }

    private void inv(int n, DoubleMatrix2D ww, DoubleMatrix2D xs, DoubleMatrix2D wwi) {
        xs.assign(Mult.mult(-1));
        int nm1 = n - 1;

        double dp3 = 0.0;

        for (int k = 0; k < n - 1; k++) {
            dp3 += xs.get(k, 0) * ww.get(k + 1, 0);
        }

        wwi.set(0, 0, 1.0 / (ww.get(0, 0) + dp3));

        for (int i = 1; i < n; i++) wwi.set(i, 0, wwi.get(0, 0) * xs.get(i - 1, 0));

        double dp4 = 0.0;

        for (int k = 0; k < n - 1; k++) {
            dp4 += xs.get(k, n - 1) * ww.get(k, n - 1);
        }

        wwi.set(n - 1, n - 1, 1.0 / (ww.get(n - 1, n - 1) + dp4));

        for (int i = 0; i < nm1; i++) {
            wwi.set(i, n - 1, wwi.get(n - 1, n - 1) * xs.get(i, n - 1));
        }

        for (int j = 1; j < n - 1; j++) {
            if (Thread.currentThread().isInterrupted()) {
                break;
            }

            int jm1 = j - 1;
            int jp1 = j + 1;

            double dp1 = 0.0;

            for (int k = 0; k <= jm1; k++) {
                dp1 += xs.get(k, j) * ww.get(k, j);
            }

            double dp2 = 0.0;

            for (int k = j; k <= n - 2; k++) {
                if (Thread.currentThread().isInterrupted()) {
                    break;
                }

                dp2 += xs.get(k, j) * ww.get(k + 1, j);
            }

            wwi.set(j, j, 1.0 / (ww.get(j, j) + dp1 + dp2));
            for (int p = 0; p <= jm1; p++) wwi.set(p, j, wwi.get(j, j) * xs.get(p, j));
            for (int p = jp1; p < n; p++) wwi.set(p, j, wwi.get(j, j) * xs.get(p - 1, j));
        }
    }

    private void zero(DoubleMatrix2D wwi) {
        for (int i = 0; i < wwi.rows(); i++) {
            for (int j = 0; j < wwi.columns(); j++) {
                wwi.set(i, j, 0.0);
            }
        }
    }

    private interface Rho {
        double get(int i, int j);
    }

    public boolean isIa() {
        return this.ia;
    }

    public void setIa(boolean ia) {
        this.ia = ia;
    }

    public boolean isIs() {
        return this.is;
    }

    public void setIs(boolean is) {
        this.is = is;
    }

    public boolean isItr() {
        return this.itr;
    }

    public void setItr(boolean itr) {
        this.itr = itr;
    }

    public boolean isIpen() {
        return this.ipen;
    }

    public void setIpen(boolean ipen) {
        this.ipen = ipen;
    }

    public double getThr() {
        return this.thr;
    }

    public int getN() {
        return this.n;
    }

    public void setN(int n) {
        if (n < 0) throw new IllegalArgumentException("Dimension >= 0: " + n);

        this.n = n;
    }

    public DoubleMatrix2D getSs() {
        return this.ss;
    }

    public void setSs(DoubleMatrix2D ss) {
        if (this.n == -1) throw new IllegalArgumentException("N (dimension) not set.");

        if (!(ss.rows() == this.n && ss.columns() == this.n)) {
            throw new IllegalArgumentException("ss not square of dimension n.");
        }

        this.ss = ss;
    }

    public Rho getRho() {
        return this.rho;
    }

    public void setRhoAllEqual(double rho) {
        this.rho = (i, j) -> rho;
    }


    public int getMaxit() {
        return this.maxit;
    }

    public void setMaxit(int maxit) {
        if (maxit <= 0) throw new IllegalArgumentException("Max iterations must be > 0: " + maxit);

        this.maxit = maxit;
    }


    public void setThr(double thr) {
        if (thr < 0) throw new IllegalArgumentException("Threshold must be >= 0: " + thr);

        this.thr = thr;
    }

}

File: docs/referenced_papers/glasso.f90
!
!           Lasso regularized covariance matrix estimate
!
!                     version (1/22/09)
!
! call lasinv(n,ss,rho,ia,is,itr,ipen,thr,maxit,ww,wwi,niter,del,jerr)
!
! Input:
!    n = dimension of matrix
!    ss(n,n) = data covariance matrix
!    rho(n,n) = regularization strength parameters for each element
!              (must be symmetric: rho(i,j)=rho(j,i))
!    ia = approximation flag
!       ai =  0 => exact solution
!       ia != 0 => Meinhausen-Buhlmann approximation
!    is = initialization flag
!       is  = 0 => cold start: initialize using ss
!       is != 0 => warm start: initialize with previous solution
!                  stored in ww and wwi (see below)
!    itr = trace flag
!       itr != 0 => trace information printed
!       itr =  0 => trace information not printed
!    ipen = diagonal penalty flag
!       ipen != 0 => diagonal is penalized
!       ipen =  0 => diagonal is not penalized
!    thr = convergence threshold: iterations stop when average absolute
!          parameter change is less than thr * ave(abs(offdiag(ss)))
!          (suggested value 1.0e-4)
!    maxit = maximum number of iterations (no effect for ia ! = 0)
!
! Output:
!    ww(n,n) = solution covariance matrix estimate (ia = 0)
!               (not used for ia != 0)
!    wwi(n,n) = solution inverse covariance matrix estimate (ia = 0)
!             = off-diagonal lasso coefficients (ia != 0)
!    niter = number of iterations
!    del = average absolute parameter change at termination
!             (not used for ia != 0)
!    jerr = memory allocation error flag
!      jerr = 0 => no error
!      jerr != 0 => memory allocation error - no output returned
!
!
      subroutine glasso(n,ss,rho,ia,is,itr,ipen,thr,maxit,ww,wwi,niter,del,jerr)
      implicit double precision(a-h, o-z)
      parameter(eps=1.0e-7)
      double precision ss(n,n),rho(n,n),ww(n,n),wwi(n,n)
      double precision, dimension (:,:), allocatable :: vv,xs
      double precision, dimension (:), allocatable :: s,x,z,ws,ro,so
      integer, dimension (:), allocatable :: mm
      nm1=n-1
      allocate(vv(1:nm1,1:nm1),stat=jerr)
      ierr=0
      if(ia.eq.0) allocate(xs(1:nm1,1:n),stat=ierr)
      jerr=jerr+ierr
      allocate(s(1:nm1),stat=ierr)
      jerr=jerr+ierr
      allocate(so(1:nm1),stat=ierr)
      jerr=jerr+ierr
      allocate(x(1:nm1),stat=ierr)
      jerr=jerr+ierr
      allocate(z(1:nm1),stat=ierr)
      jerr=jerr+ierr
      allocate(mm(1:nm1),stat=ierr)
      jerr=jerr+ierr
      allocate(ro(1:nm1),stat=ierr)
      jerr=jerr+ierr
      if(ia .ne. 0)goto 10021
      allocate(ws(1:n),stat=ierr)
      jerr=jerr+ierr
10021 continue
      if(jerr.ne.0) return
      shr=0.0
10030 do 10031 j=1,n
10040 do 10041 k=1,n
      if(j.eq.k)goto 10041
      shr=shr+abs(ss(j,k))
10041 continue
10042 continue
10031 continue
10032 continue
      if(shr .ne. 0.0)goto 10061
      ww=0.0
      wwi=0.0
10070 do 10071 j=1,n
      if(ipen .ne. 0)goto 10091
      ww(j,j)=ss(j,j)
      goto 10101
10091 continue
      ww(j,j)=ss(j,j)+rho(j,j)
10101 continue
10081 continue
      wwi(j,j)=1.0/max(ww(j,j),eps)
10071 continue
10072 continue
      return
10061 continue
      shr=thr*shr/nm1
      if(ia .eq. 0)goto 10121
      if(is.eq.0) wwi=0.0
10130 do 10131 m=1,n
      call setup(m,n,ss,rho,ss,vv,s,ro)
      l=0
10140 do 10141 j=1,n
      if(j.eq.m)goto 10141
      l=l+1
      x(l)=wwi(j,m)
10141 continue
10142 continue
      call lasso(ro,nm1,vv,s,shr/n,x,z,mm)
      l=0
10150 do 10151 j=1,n
      if(j.eq.m)goto 10151
      l=l+1
      wwi(j,m)=x(l)
10151 continue
10152 continue
10131 continue
10132 continue
      niter=1
      return
10121 continue
      if(is .ne. 0)goto 10171
      ww=ss
      xs=0.0
      goto 10181
10171 continue
10190 do 10191 j=1,n
      xjj=-wwi(j,j)
      l=0
10200 do 10201 k=1,n
      if(k.eq.j)goto 10201
      l=l+1
      xs(l,j)=wwi(k,j)/xjj
10201 continue
10202 continue
10191 continue
10192 continue
10181 continue
10161 continue
10210 do 10211 j=1,n
      if(ipen .ne. 0)goto 10231
      ww(j,j)=ss(j,j)
      goto 10241
10231 continue
      ww(j,j)=ss(j,j)+rho(j,j)
10241 continue
10221 continue
10211 continue
10212 continue
      niter=0
10250 continue
10251 continue
      dlx=0.0
10260 do 10261 m=1,n
      if(itr .eq. 0)goto 10281
!      write(6,10290)m
10290 format ('outer loop, m =',i10)
10281 continue
      x=xs(:,m)
      ws=ww(:,m)
      call setup(m,n,ss,rho,ww,vv,s,ro)
      so=s
      call lasso(ro,nm1,vv,s,shr/sum(abs(vv)),x,z,mm)
      l=0
10300 do 10301 j=1,n
      if(j.eq.m)goto 10301
      l=l+1
      ww(j,m)=so(l)-s(l)
      ww(m,j)=ww(j,m)
10301 continue
10302 continue
      dlx=max(dlx,sum(abs(ww(:,m)-ws)))
      xs(:,m)=x
10261 continue
10262 continue
      niter=niter+1
      if(niter.ge.maxit)goto 10252
      if(dlx.lt.shr)goto 10252
      goto 10251
10252 continue
      del=dlx/nm1
      call inv(n,ww,xs,wwi)
      return
      end
      subroutine setup(m,n,ss,rho,ww,vv,s,r)
      implicit double precision(a-h, o-z)
      double precision ss(n,n),rho(n,n),ww(n,n),vv(n-1,n-1),s(n-1),r(n-1)
      l=0
10310 do 10311 j=1,n
      if(j.eq.m)goto 10311
      l=l+1
      r(l)=rho(j,m)
      s(l)=ss(j,m)
      i=0
10320 do 10321 k=1,n
      if(k.eq.m)goto 10321
      i=i+1
      vv(i,l)=ww(k,j)
10321 continue
10322 continue
10311 continue
10312 continue
      return
      end
      subroutine lasso(rho,n,vv,s,thr,x,z,mm)
      implicit double precision(a-h, o-z)
      double precision rho(n),vv(n,n),s(n),x(n),z(n)
      integer mm(n)
      call fatmul(2,n,vv,x,s,z,mm)
10330 continue
10331 continue
      dlx=0.0
10340 do 10341 j=1,n
      xj=x(j)
      x(j)=0.0
      t=s(j)+vv(j,j)*xj
      if (abs(t)-rho(j).gt.0.0) x(j)=sign(abs(t)-rho(j),t)/vv(j,j)
      if(x(j).eq.xj)goto 10341
      del=x(j)-xj
      dlx=max(dlx,abs(del))
      s=s-del*vv(:,j)
10341 continue
10342 continue
      if(dlx.lt.thr)goto 10332
      goto 10331
10332 continue
      return
      end
      subroutine fatmul(it,n,vv,x,s,z,m)
      implicit double precision(a-h, o-z)
      parameter(fac=0.2)
      double precision vv(n,n),x(n),s(n),z(n)
      integer m(n)
      l=0
10350 do 10351 j=1,n
      if(x(j).eq.0.0)goto 10351
      l=l+1
      m(l)=j
      z(l)=x(j)
10351 continue
10352 continue
      if(l .le. int(fac*n))goto 10371
      if(it .ne. 1)goto 10391
      s=matmul(vv,x)
      goto 10401
10391 continue
      s=s-matmul(x,vv)
10401 continue
10381 continue
      goto 10361
10371 if(it .ne. 1)goto 10411
10420 do 10421 j=1,n
      s(j)=dot_product(vv(j,m(1:l)),z(1:l))
10421 continue
10422 continue
      goto 10431
10411 continue
10440 do 10441 j=1,n
      s(j)=s(j)-dot_product(vv(m(1:l),j),z(1:l))
10441 continue
10442 continue
10431 continue
10361 continue
      return
      end
      subroutine inv(n,ww,xs,wwi)
      implicit double precision(a-h, o-z)
      double precision ww(n,n),xs(n-1,n),wwi(n,n)
      nm1=n-1
      xs=-xs
      wwi(1,1)=1.0/(ww(1,1)+dot_product(xs(:,1),ww(2:n,1)))
      wwi(2:n,1)=wwi(1,1)*xs(:,1)
      wwi(n,n)=1.0/(ww(n,n)+dot_product(xs(:,n),ww(1:nm1,n)))
      wwi(1:nm1,n)=wwi(n,n)*xs(:,n)
10450 do 10451 j=2,nm1
      jm1=j-1
      jp1=j+1
      wwi(j,j)=1.0/(ww(j,j)+dot_product(xs(1:jm1,j),ww(1:jm1,j))  +dot_product(xs(j:nm1,j),ww(jp1:n,j)))
      wwi(1:jm1,j)=wwi(j,j)*xs(1:jm1,j)
      wwi(jp1:n,j)=wwi(j,j)*xs(j:nm1,j)
10451 continue
10452 continue
      return
      end




! Output from Public domain Ratfor, version 1.0
!     coded by R tibshirani dec 2008

      subroutine glassopath(beta,what,jerrs,rholist, nrho,n,ss,rho,ia,itr,ipen,thr,maxit,ww,wwi,niter,del,jerr)
      implicit double precision(a-h, o-z)

      integer nrho,n,jerrs(nrho)
      double precision rholist(nrho),beta(n,n,nrho),what(n,n,nrho)
      double precision ss(n,n),rho(n,n),ww(n,n),wwi(n,n)
      is=0
      do 23000 j=1,n
      do 23002 k=1,n
      rho(j,k)=rholist(nrho)
23002 continue
23003 continue
23000 continue
23001 continue
      call glasso(n,ss,rho,ia,is,itr,ipen,thr,maxit,ww,wwi,niter,del,jerr)
      jerrs(1)=jerr
      do 23004 j=1,n
      do 23006 k=1,n
      beta(j,k,nrho)=wwi(j,k)
      what(j,k,nrho)=ww(j,k)
23006 continue
23007 continue
23004 continue
23005 continue
      is=1
      do 23008 i =(nrho-1), 1,-1
      do 23010 j=1,n
      do 23012 k=1,n
      rho(j,k)=rholist(i)
23012 continue
23013 continue
23010 continue
23011 continue
      if(itr.eq.1) write(6,*) "rho=",rholist(i)
      call glasso(n,ss,rho,ia,is,itr,ipen,thr,maxit,ww,wwi,niter,del,jerr)
      jerrs(i)=jerr
      do 23014 j=1,n
      do 23016 k=1,n
      beta(j,k,i)=wwi(j,k)
      what(j,k,i)=ww(j,k)
23016 continue
23017 continue
23014 continue
23015 continue
23008 continue
23009 continue
      return
      end
